File: src/main/scala/org/winlogon/whisperchat/WhisperChatPlugin.scala
package org.winlogon.whisperchat

import org.bukkit.Bukkit
import org.bukkit.command.{Command, CommandExecutor, CommandSender}
import org.bukkit.entity.Player
import org.bukkit.event.{EventHandler, EventPriority, Listener}
import org.bukkit.event.player.AsyncPlayerChatEvent
import org.bukkit.plugin.java.JavaPlugin

import java.util.UUID
import scala.collection.concurrent.TrieMap

class WhisperChatPlugin extends JavaPlugin {
  private val activeDMs = TrieMap.empty[UUID, UUID]
  private val lastSenders = TrieMap.empty[UUID, UUID]
  private var isFolia: Boolean = false

  override def onEnable(): Unit = {
    isFolia = checkFolia()
    getCommand("dm").setExecutor(new DMCommandExecutor(this))
    getCommand("r").setExecutor(new ReplyCommandExecutor(this))
    Bukkit.getPluginManager.registerEvents(new ChatListener(this), this)
  }

  private def checkFolia(): Boolean = {
    try {
      Class.forName("io.papermc.paper.threadedregions.RegionizedServer")
      true
    } catch {
      case _: ClassNotFoundException => false
    }
  }

  def getActiveDMs: TrieMap[UUID, UUID] = activeDMs
  def getLastSenders: TrieMap[UUID, UUID] = lastSenders
  def isServerFolia: Boolean = isFolia
}

class DMCommandExecutor(plugin: WhisperChatPlugin) extends CommandExecutor {
  private val activeDMs = plugin.getActiveDMs
  private val lastSenders = plugin.getLastSenders

  override def onCommand(sender: CommandSender, command: Command, label: String, args: Array[String]): Boolean = {
    if (!sender.isInstanceOf[Player]) {
      sender.sendMessage("§cOnly players can use this command.")
      return true
    }

    val player = sender.asInstanceOf[Player]

    if (args.isEmpty) {
      player.sendMessage("§7Usage: §3/dm §2<username|leave>")
      return true
    }

    if (args(0).equalsIgnoreCase("leave")) {
      activeDMs.remove(player.getUniqueId)
      player.sendMessage("§7You have §3left§7 the DM. Chat is now §2global§7.")
      return true
    }

    val target = Bukkit.getPlayer(args(0))
    if (target == null) {
      player.sendMessage(s"§cError: §7Player §3${args(0)}§7 not found or offline.")
      return true
    }

    if (target == player) {
      player.sendMessage("§cnotepad exists mate")
      return true
    }

    activeDMs.put(player.getUniqueId, target.getUniqueId)
    lastSenders.put(target.getUniqueId, player.getUniqueId)
    player.sendMessage(s"§7Now DMing §3${target.getName}§7. Type normally in chat to send private messages.")
    true
  }
}

class ReplyCommandExecutor(plugin: WhisperChatPlugin) extends CommandExecutor {
  private val lastSenders = plugin.getLastSenders

  override def onCommand(sender: CommandSender, command: Command, label: String, args: Array[String]): Boolean = {
    if (!sender.isInstanceOf[Player]) {
      sender.sendMessage("§cOnly players can use this command.")
      return true
    }

    val player = sender.asInstanceOf[Player]

    if (args.isEmpty) {
      player.sendMessage("§7Usage: §3/r §2<message>")
      return true
    }

    val lastSender = lastSenders.get(player.getUniqueId)

    lastSender match {
      case Some(lastSenderUuid) =>
        val target = Bukkit.getPlayer(lastSenderUuid)
        if (target == null || !target.isOnline) {
          player.sendMessage("§7The player you're replying to is §3offline.")
          lastSenders.remove(player.getUniqueId)
          true
        } else {
          val message = args.mkString(" ")
          target.sendMessage(s"§7${player.getName} whispers to you: §f$message")
          player.sendMessage(s"§7You whisper to §6${target.getName}§7: §f$message")
          lastSenders.put(target.getUniqueId, player.getUniqueId)
          true
        }
      case None =>
        player.sendMessage("§cYou have no one to reply to.")
        true
    }
  }
}

class ChatListener(plugin: WhisperChatPlugin) extends Listener {
  private val activeDMs = plugin.getActiveDMs
  private val lastSenders = plugin.getLastSenders
  private val isFolia = plugin.isServerFolia

  @EventHandler(priority = EventPriority.LOWEST)
  def onChat(event: AsyncPlayerChatEvent): Unit = {
    val player = event.getPlayer
    activeDMs.get(player.getUniqueId).foreach { targetUuid =>
      event.setCancelled(true)
      val message = event.getMessage
      val senderUuid = player.getUniqueId

      // Create a proper Java Runnable instance
      val runnable = new Runnable {
        override def run(): Unit = {
          val sender = Bukkit.getPlayer(senderUuid)
          if (sender == null || !sender.isOnline) return

          val target = Bukkit.getPlayer(targetUuid)
          if (target == null || !target.isOnline) {
            sender.sendMessage("§cPlayer is offline. Returning to global chat.")
            activeDMs.remove(senderUuid)
          } else {
            target.sendMessage(s"§7${sender.getName} whispers to you: §f$message")
            sender.sendMessage(s"§7You whisper to §6${target.getName}§7: §f$message")
            lastSenders.put(target.getUniqueId, senderUuid)
          }
        }
      }

      if (isFolia) {
        player.getScheduler.run(plugin, _ => runnable.run(), null)
      } else {
        Bukkit.getScheduler.runTask(plugin, runnable)
      }
    }
  }
}

File: src/main/scala/org/winlogon/whisperchat/WhisperChatPlugin.kt
package org.winlogon.whisperchat

import dev.jorel.commandapi.CommandAPI
import dev.jorel.commandapi.CommandAPICommand
import dev.jorel.commandapi.arguments.*
import org.bukkit.Bukkit
import org.bukkit.configuration.file.FileConfiguration
import org.bukkit.entity.Player
import org.bukkit.event.EventHandler
import org.bukkit.event.EventPriority
import org.bukkit.event.Listener
import org.bukkit.event.player.AsyncPlayerChatEvent
import org.bukkit.plugin.java.JavaPlugin
import java.util.*
import java.util.concurrent.ConcurrentHashMap

class WhisperChatPlugin : JavaPlugin() {
    private val activeDMs = ConcurrentHashMap<UUID, UUID>()
    private val dmSessions = ConcurrentHashMap<UUID, MutableSet<UUID>>()
    private val lastSenders = ConcurrentHashMap<UUID, UUID>()
    private var isFolia = false
    private lateinit var config: FileConfiguration

    override fun onEnable() {
        saveDefaultConfig()
        config = config
        isFolia = checkFolia()
        CommandAPI.onEnable(this)
        registerCommands()
        server.pluginManager.registerEvents(ChatListener(this), this)
    }

    private fun checkFolia(): Boolean {
        return try {
            Class.forName("io.papermc.paper.threadedregions.RegionizedServer")
            true
        } catch (e: ClassNotFoundException) {
            false
        }
    }

    private fun registerCommands() {
        // DM Command
        CommandAPICommand("dm")
            .withSubcommand(
                CommandAPICommand("start")
                    .withArguments(PlayerArgument("target"))
                    .executesPlayer { player, args ->
                        val target = args[0] as Player
                        handleDMStart(player, target)
                    }
            )
            .withSubcommand(
                CommandAPICommand("switch")
                    .withArguments(PlayerArgument("target"))
                    .executesPlayer { player, args ->
                        val target = args[0] as Player
                        handleDMSwitch(player, target)
                    }
            )
            .withSubcommand(
                CommandAPICommand("list")
                    .executesPlayer { player, _ ->
                        handleDMList(player)
                    }
            )
            .withSubcommand(
                CommandAPICommand("leave")
                    .executesPlayer { player, _ ->
                        handleDMLeave(player)
                    }
            )
            .withSubcommand(
                CommandAPICommand("help")
                    .executesPlayer { player, _ ->
                        sendHelp(player)
                    }
            )
            .register()

        // Reply Command
        CommandAPICommand("r")
            .withArguments(GreedyStringArgument("message"))
            .executesPlayer { player, args ->
                val message = args[0] as String
                handleReply(player, message)
            }
            .register()

        // Whisper/Msg Commands
        val whisperCmd = CommandAPICommand("w")
            .withArguments(PlayerArgument("target"))
            .withArguments(GreedyStringArgument("message"))
            .executesPlayer { player, args ->
                val target = args[0] as Player
                val message = args[1] as String
                sendFormattedMessage(player, target, message, "whisper")
            }
        
        CommandAPICommand("msg")
            .withArguments(PlayerArgument("target"))
            .withArguments(GreedyStringArgument("message"))
            .executesPlayer(whisperCmd.executor)
            .register()
    }

    private fun handleDMStart(player: Player, target: Player) {
        if (player == target) {
            player.sendMessage(config.getString("messages.self-whisper-error"))
            return
        }

        dmSessions.compute(player.uniqueId) { _, sessions ->
            (sessions ?: mutableSetOf()).apply { add(target.uniqueId) }
        }
        activeDMs[player.uniqueId] = target.uniqueId
        lastSenders[target.uniqueId] = player.uniqueId
        
        player.sendMessage(config.getString("messages.dm-start")
            ?.replace("{target}", target.name))
    }

    private fun handleDMSwitch(player: Player, target: Player) {
        val sessions = dmSessions[player.uniqueId] ?: run {
            player.sendMessage(config.getString("messages.no-dm-sessions"))
            return
        }

        if (target.uniqueId in sessions) {
            activeDMs[player.uniqueId] = target.uniqueId
            player.sendMessage(config.getString("messages.dm-switch")
                ?.replace("{target}", target.name))
        } else {
            player.sendMessage(config.getString("messages.invalid-dm-target"))
        }
    }

    private fun handleDMList(player: Player) {
        val sessions = dmSessions[player.uniqueId] ?: run {
            player.sendMessage(config.getString("messages.no-dm-sessions"))
            return
        }

        val onlineTargets = sessions.mapNotNull { Bukkit.getPlayer(it)?.name }
        if (onlineTargets.isEmpty()) {
            player.sendMessage(config.getString("messages.no-active-dms"))
            return
        }

        player.sendMessage(config.getString("messages.dm-list-header"))
        onlineTargets.forEach { target ->
            player.sendMessage(config.getString("messages.dm-list-item")
                ?.replace("{target}", target))
        }
    }

    private fun handleDMLeave(player: Player) {
        val targetId = activeDMs[player.uniqueId] ?: run {
            player.sendMessage(config.getString("messages.not-in-dm"))
            return
        }

        activeDMs.remove(player.uniqueId)
        dmSessions.computeIfPresent(player.uniqueId) { _, sessions ->
            sessions.apply { remove(targetId) }
        }
        player.sendMessage(config.getString("messages.dm-left")
            ?.replace("{target}", Bukkit.getPlayer(targetId)?.name ?: "Unknown"))
    }

    private fun handleReply(player: Player, message: String) {
        val lastSenderId = lastSenders[player.uniqueId] ?: run {
            player.sendMessage(config.getString("messages.no-reply-target"))
            return
        }

        val target = Bukkit.getPlayer(lastSenderId) ?: run {
            lastSenders.remove(player.uniqueId)
            player.sendMessage(config.getString("messages.target-offline"))
            return
        }

        sendFormattedMessage(player, target, message, "reply")
        lastSenders[target.uniqueId] = player.uniqueId
    }

    fun sendFormattedMessage(sender: Player, receiver: Player, message: String, type: String) {
        val formatKey = when (type) {
            "whisper" -> "formats.whisper"
            "reply" -> "formats.reply"
            else -> "formats.default"
        }

        val format = config.getString(formatKey, "&7[{type}] {sender} &7-> &6{receiver}&7: &f{message}")
            .replace("{type}", type.uppercase())
            .replace("{sender}", sender.name)
            .replace("{receiver}", receiver.name)
            .replace("{message}", message)

        sender.sendMessage(format)
        receiver.sendMessage(format)
    }

    private fun sendHelp(player: Player) {
        config.getStringList("messages.help").forEach(player::sendMessage)
    }

    override fun onDisable() {
        CommandAPI.onDisable()
    }

    inner class ChatListener : Listener {
        @EventHandler(priority = EventPriority.LOWEST)
        fun onChat(event: AsyncPlayerChatEvent) {
            val player = event.player
            val targetId = activeDMs[player.uniqueId] ?: return
            event.isCancelled = true

            val message = event.message
            val runnable = Runnable {
                val target = Bukkit.getPlayer(targetId) ?: run {
                    player.sendMessage(config.getString("messages.target-offline"))
                    activeDMs.remove(player.uniqueId)
                    dmSessions.computeIfPresent(player.uniqueId) { _, sessions ->
                        sessions.apply { remove(targetId) }
                    }
                    return@Runnable
                }

                sendFormattedMessage(player, target, message, "dm")
                lastSenders[target.uniqueId] = player.uniqueId
            }

            if (isFolia) {
                player.scheduler.run(this@WhisperChatPlugin, { _ -> runnable.run() }, null)
            } else {
                Bukkit.getScheduler().runTask(this@WhisperChatPlugin, runnable)
            }
        }
    }
}
